src/runtime/graph.cc

```diff
@@ -1542,81 +1874,265 @@ size_t dp_state_hash(Graph const *graph,
+      data_parallel_view.stride[0] = 1;
+      data_parallel_view.start_device_id = 0;
+      for (auto const &node : curr_best_graph->inEdges) {
+        curr_optimal_views[node.first] = data_parallel_view;
+      }
+    } else {
+      // Main step to optimize the PCG of an FFModel
+      model->graph_optimize(model->config.search_budget,
+                            model->config.only_data_parallel,
+                            curr_best_graph,
+                            curr_optimal_views,
+                            perform_memory_search,
+                            MemoryOptimConfig{lambda.first},
+                            lambda.second);
+    }
+    // Return the best result of the current search
+    return std::make_pair(std::move(curr_best_graph), curr_optimal_views);
+  };
+
+  /**
+   * @brief Analyze the per-device memory cost and compare with the memory
+   * threshold of each device.
+   */
+  auto is_valid_strategy =
+      [&](std::vector<std::pair<float, MemorySearchResult>> &lambdas_results,
+          Graph *curr_graph,
+          std::unordered_map<Node, MachineView> &curr_views) {
+        std::cout << "try to check valid for lambda "
+                  << lambdas_results.back().first << std::endl;
+
+        // Analyze the strategy and update max_per_device_mem_all_deivces in the
+        // lambda_result.
+        std::unordered_map<int, float> device_to_mem{};
+        for (auto const &view : curr_views) {
+          CostMetrics op_cost = cached_simulator->measure_operator_cost(
+              view.first.ptr, view.second);
+          float node_mem_as_mb = op_cost.total_memory_in_mb();
+
+          for (const auto d_id : view.second.device_ids()) {
+            if (device_to_mem.find(d_id) == device_to_mem.end()) {
+              device_to_mem.emplace(std::make_pair(d_id, node_mem_as_mb));
+            } else {
+              device_to_mem[d_id] += node_mem_as_mb;
+            }
+          }
+        }
+
+        float max_per_device_mem = 0.0;
+        float total_device_mem = 0.0;
+        for (const auto &d : device_to_mem) {
+          std::cout << "d_id: " << d.first << ", mem: " << d.second
+                    << std::endl;
+          total_device_mem += d.second;
+          if (d.second > max_per_device_mem) {
+            max_per_device_mem = d.second;
+          }
+        }
+
+        lambdas_results.back().second.max_per_device_mem_all_deivces =
+            max_per_device_mem;
+
+        std::cout
+            << "max_per_device_mem: "
+            << lambdas_results.back().second.max_per_device_mem_all_deivces
+            << ", total_device_mem: " << total_device_mem << std::endl;
+
+        if (max_per_device_mem >= memory_threshold) {
+          return false;
+        }
+        return true;
+      };
+
+  // Be optimistic
+  lambdas.emplace_back(std::make_pair(1.0, MemorySearchResult{}));
+  auto try_result = try_one_lambda(lambdas.back());
```

Comment:
Adding on to the need for a refactor -- maybe `auto` can be specified to the actual type? I'm not sure what `.first` and `.second` are supposed to be.

Reply:
I have factored this lambda out. Hopefully, it became more clear now.